In [1]:
import pandas as pd 
import numpy as np
from sklearn import preprocessing
from sklearn import tree
from sklearn.ensemble import RandomForestClassifier


# 1. Building Decision Tree for Attrition Dataset

In [2]:
df1 = pd.read_csv("general_data.csv")

In [3]:
df1.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [4]:
df1.isna().sum()

Age                         0
Attrition                   0
BusinessTravel              0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeID                  0
Gender                      0
JobLevel                    0
JobRole                     0
MaritalStatus               0
MonthlyIncome               0
NumCompaniesWorked         19
Over18                      0
PercentSalaryHike           0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           9
TrainingTimesLastYear       0
YearsAtCompany              0
YearsSinceLastPromotion     0
YearsWithCurrManager        0
dtype: int64

In [9]:
df1["NumCompaniesWorked"]= np.where(df1["NumCompaniesWorked"].isnull(),df1["NumCompaniesWorked"].mean(), df1["NumCompaniesWorked"])
df1["TotalWorkingYears"]= np.where(df1["TotalWorkingYears"].isnull(),df1["TotalWorkingYears"].mean(), df1["TotalWorkingYears"])


In [10]:
df1.isna().sum()

Age                        0
Attrition                  0
BusinessTravel             0
Department                 0
DistanceFromHome           0
Education                  0
EducationField             0
EmployeeCount              0
EmployeeID                 0
Gender                     0
JobLevel                   0
JobRole                    0
MaritalStatus              0
MonthlyIncome              0
NumCompaniesWorked         0
Over18                     0
PercentSalaryHike          0
StandardHours              0
StockOptionLevel           0
TotalWorkingYears          0
TrainingTimesLastYear      0
YearsAtCompany             0
YearsSinceLastPromotion    0
YearsWithCurrManager       0
dtype: int64

In [11]:
le = preprocessing.LabelEncoder()

In [12]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Data columns (total 24 columns):
Age                        4410 non-null int64
Attrition                  4410 non-null object
BusinessTravel             4410 non-null object
Department                 4410 non-null object
DistanceFromHome           4410 non-null int64
Education                  4410 non-null int64
EducationField             4410 non-null object
EmployeeCount              4410 non-null int64
EmployeeID                 4410 non-null int64
Gender                     4410 non-null object
JobLevel                   4410 non-null int64
JobRole                    4410 non-null object
MaritalStatus              4410 non-null object
MonthlyIncome              4410 non-null int64
NumCompaniesWorked         4410 non-null float64
Over18                     4410 non-null object
PercentSalaryHike          4410 non-null int64
StandardHours              4410 non-null int64
StockOptionLevel           4410 non-n

In [13]:
df1['Attrition'] = le.fit_transform(df1.Attrition)
df1['BusinessTravel'] = le.fit_transform(df1.BusinessTravel)
df1['Department'] = le.fit_transform(df1.Department)
df1['EducationField'] = le.fit_transform(df1.EducationField)
df1['Gender'] = le.fit_transform(df1.Gender)
df1['JobRole'] = le.fit_transform(df1.JobRole)
df1['MaritalStatus'] = le.fit_transform(df1.MaritalStatus)
df1['Over18'] = le.fit_transform(df1.Over18)

In [15]:
model1 = RandomForestClassifier(n_estimators = 2000, max_features = 2, oob_score = True)

In [16]:
df1.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EmployeeCount', 'EmployeeID', 'Gender',
       'JobLevel', 'JobRole', 'MaritalStatus', 'MonthlyIncome',
       'NumCompaniesWorked', 'Over18', 'PercentSalaryHike', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager'],
      dtype='object')

In [19]:
df1 = df1.drop(['EmployeeCount',  'EmployeeID', 'StandardHours'],axis =1 )

In [20]:
features1 = ['Age',  'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField',  'Gender',
       'JobLevel', 'JobRole', 'MaritalStatus', 'MonthlyIncome',
       'NumCompaniesWorked', 'Over18', 'PercentSalaryHike', 
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager']

In [21]:
model1.fit(X = df1[features1], y = df1.Attrition)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features=2, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=2000, n_jobs=None,
            oob_score=True, random_state=None, verbose=0, warm_start=False)

In [22]:
print("OOB Accuracy:")
print(model1.oob_score_)

OOB Accuracy:
1.0


In [24]:
for feature , imp in zip(features1,model1.feature_importances_):
    print(feature,"---------------------", imp)

Age --------------------- 0.09646573820000517
BusinessTravel --------------------- 0.0280096099424294
Department --------------------- 0.025948584683808198
DistanceFromHome --------------------- 0.06941460604057913
Education --------------------- 0.04091034988258575
EducationField --------------------- 0.04128487634223856
Gender --------------------- 0.018201352373424916
JobLevel --------------------- 0.038329737282449046
JobRole --------------------- 0.05615912952299589
MaritalStatus --------------------- 0.039566393766740995
MonthlyIncome --------------------- 0.09373903291989728
NumCompaniesWorked --------------------- 0.05602099442733852
Over18 --------------------- 0.0
PercentSalaryHike --------------------- 0.06558464126045117
StockOptionLevel --------------------- 0.03412679155204756
TotalWorkingYears --------------------- 0.08565762428227214
TrainingTimesLastYear --------------------- 0.04493655007732865
YearsAtCompany --------------------- 0.0679933755837414
YearsSinceLastProm

## Inference 
### Most Important Features are Age, MonthlyIncome, TotalWorkingYears, DistanceFromHome, PercentSalaryHike, YearsAtCompany , JobRole, NumCompaniesWorked and YearsWithCurrManager

In [25]:
Features2 = ['Age', 'MonthlyIncome', 'TotalWorkingYears', 'DistanceFromHome', 'PercentSalaryHike', 'YearsAtCompany' , 'JobRole', 'NumCompaniesWorked' ,  'YearsWithCurrManager']

In [26]:
model_dt = tree.DecisionTreeClassifier(max_depth = 12)

In [28]:
model_dt.fit(X = df1[Features2], y=df1.Attrition)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=12,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [29]:
model_dt.score(X = df1[Features2], y=df1.Attrition)

0.9582766439909297

## Inference
 Score is 95.82%. So, we can say that Age, MonthlyIncome, TotalWorkingYears, DistanceFromHome, PercentSalaryHike, YearsAtCompany, JobRole, NumCompaniesWorked and YearswithCurrManager features play important role in attrition level in the company. 

# 2. Building Decision Tree for Bank Loan Dataset

In [30]:
df2 = pd.read_excel("Bank_Personal_Loan_Modelling.xlsx", sheet_name="Data")
df2.head()

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [31]:
df2.isna().sum()

ID                    0
Age                   0
Experience            0
Income                0
ZIP Code              0
Family                0
CCAvg                 0
Education             0
Mortgage              0
Personal Loan         0
Securities Account    0
CD Account            0
Online                0
CreditCard            0
dtype: int64

In [33]:
df2.dtypes

ID                      int64
Age                     int64
Experience              int64
Income                  int64
ZIP Code                int64
Family                  int64
CCAvg                 float64
Education               int64
Mortgage                int64
Personal Loan           int64
Securities Account      int64
CD Account              int64
Online                  int64
CreditCard              int64
dtype: object

In [34]:
model_bank = RandomForestClassifier(n_estimators = 2000, max_features = 2 , oob_score = True)

In [37]:
df2 = df2.drop(['ID','ZIP Code'],axis=1)

In [38]:
df2.columns

Index(['Age', 'Experience', 'Income', 'Family', 'CCAvg', 'Education',
       'Mortgage', 'Personal Loan', 'Securities Account', 'CD Account',
       'Online', 'CreditCard'],
      dtype='object')

In [39]:
features2 = ['Age', 'Experience', 'Income', 'Family', 'CCAvg', 'Education',
       'Mortgage',  'Securities Account', 'CD Account',
       'Online', 'CreditCard']

In [42]:
model_bank.fit(X = df2[features2], y = df2['Personal Loan'])

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features=2, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=2000, n_jobs=None,
            oob_score=True, random_state=None, verbose=0, warm_start=False)

In [43]:
print("OOB Accuracy:")
print(model_bank.oob_score_)

OOB Accuracy:
0.9876


In [44]:
for feature , imp in zip(features2,model_bank.feature_importances_):
    print(feature,"---------------------", imp)

Age --------------------- 0.04441378639742495
Experience --------------------- 0.04415589006332511
Income --------------------- 0.35159681105515816
Family --------------------- 0.09755315936459605
CCAvg --------------------- 0.18186289907203046
Education --------------------- 0.15846445355104358
Mortgage --------------------- 0.043692582193659395
Securities Account --------------------- 0.005490144255549737
CD Account --------------------- 0.054580060846186426
Online --------------------- 0.008359967948537937
CreditCard --------------------- 0.009830245252487697


## Inference 
    Most Important Features are Age, CCAvg, Education, Family, CD Account and Mortage

In [45]:
bank_features = ['Family', 'CCAvg', 'Education', 'CD Account','Age','Mortgage']

In [46]:
model_bank1 = tree.DecisionTreeClassifier(max_depth = 12)

In [47]:
model_bank1.fit(X = df2[bank_features], y = df2['Personal Loan'])

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=12,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [48]:
model_bank1.score(X = df2[bank_features], y = df2['Personal Loan'])

0.9868

## Inference
    With score of 98.68% we can say that Income, CCAvg, Education, Family , CD Account features play important role in deciding whether the person will get a loan or not.

# 3. Building Decision Tree for Titanic Dataset

In [50]:
df3_train = pd.read_csv("train (1).csv")

In [51]:
df3_test = pd.read_csv("test (1).csv")

In [53]:
df3_train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [54]:
df3_test.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Embarked  
0  34.5      0      0   330911   7.8292        Q  
1  47.0      1      0   363272   7.0000        S  
2  62.0      0      0   240276   9.6875        Q  
3  27.0      0      0   315154   8.6625        S  
4  22.0      1      1  3101298  12.2875        S

In [55]:
df3_train.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64

In [56]:
df3_test.isna().sum()

PassengerId    0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

In [60]:
label_enc = preprocessing.LabelEncoder()

In [61]:
train_gender = label_enc.fit_transform(df3_train.Sex)
test_gender = label_enc.fit_transform(df3_test.Sex)

In [63]:
features3 = pd.DataFrame([df3_train.Age, df3_train.Fare, train_gender]).T

In [65]:
model3 = tree.DecisionTreeClassifier(max_depth = 6)
model3.fit(X = features3, y = df3_train.Survived)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=6,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [66]:
model3.score(X = features3, y = df3_train.Survived)

0.8267716535433071

In [67]:
features4 = pd.DataFrame([df3_test.Age, df3_test.Fare, test_gender]).T


In [68]:
pred = model3.predict(X = features4)

In [69]:
pred_output = pd.DataFrame({"Passenger_Id": df3_test.PassengerId, "Survived" : pred})

In [70]:
pred_output.head()

Passenger_Id  Survived
0           892         0
1           893         1
2           894         0
3           895         0
4           896         1

In [71]:
with open('DecisionTree1.dot','w') as file:
    file = tree.export_graphviz(model3, feature_names=['AGE','FARE','GENDER'], out_file= file)